sdk/core/azure-core/inc/azure/core/paged_response.hpp

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Provides support for paged responses.
+ */
+
+#pragma once
+
+#include <cstdlib>
+#include <string>
+
+#include "azure/core/context.hpp"
+#include "azure/core/http/raw_response.hpp"
+#include "azure/core/nullable.hpp"
+
+namespace Azure { namespace Core {
+
+  /**
+   * @brief Defines the base type and behavior for a paged response.
+   *
+   * @remark The template is using for making static-inheritance.
+   *
+   * @remark T classes must implement the way to get and move to the next page.
+   *
+   * @tparam T A class type for static-inheritance.
+   */
+  template <class T> class PagedResponse {
+  private:
+    // The field used to check when the end of the response is reached. All responses from a service
+    // will always come with a payload that represents a page. The page might or might not contain
+    // elements in the page. `m_hasPage` is then turned to `false` once `MoveToNextPage` is called
+    // on the last page.
+    bool m_hasPage = true;
+
+  protected:
+    PagedResponse() = default;
+    PagedResponse(PagedResponse&&) = default;
+    PagedResponse& operator=(PagedResponse&&) = default;
+
+  public:
+    /**
+     * @brief Defines the token used to fetch the current page.
+     *
+     */
+    std::string CurrentPageToken;
+
+    /**
+     * @brief Defines the token for getting a next page.
+     *
+     * @remark If there is not a next page, this field becomes an empty string.
+     *
+     * @remark Assumes all services will include NextPageToken in the payload, either null or empty
+     * for last page or a value for getting the next page.
+     *
+     */
+    Azure::Nullable<std::string> NextPageToken;
+
+    /**
+     * @brief The HTTP response returned by the service.
+     *
+     */
+    std::unique_ptr<Azure::Core::Http::RawResponse> RawResponse;
+
+    /**
+     * @brief Check if the page has gone after the last page to the end sentinel.
+     *
+     */
+    bool HasPage() const { return m_hasPage; }
+
+    /**
+     * @brief Get the next page.
```

Comment:
We don't return anything, so I recommend against using `Get` here.

```suggestion
     * @brief Move to the next page of the response.
```